ecs/pod_definition_manager.go

```diff
@@ -0,0 +1,152 @@
+}
+
+// NewBasicPodDefinitionManagerOptions returns new uninitialized options to
+// create a basic pod definition manager.
+func NewBasicPodDefinitionManagerOptions() *BasicPodDefinitionManagerOptions {
+	return &BasicPodDefinitionManagerOptions{}
+}
+
+// SetClient sets the client the pod uses to communicate with ECS.
+func (o *BasicPodDefinitionManagerOptions) SetClient(c cocoa.ECSClient) *BasicPodDefinitionManagerOptions {
+	o.Client = c
+	return o
+}
+
+// SetVault sets the vault that the pod uses to manage secrets.
+func (o *BasicPodDefinitionManagerOptions) SetVault(v cocoa.Vault) *BasicPodDefinitionManagerOptions {
+	o.Vault = v
+	return o
+}
+
+// SetCache sets the cache used to track pod definitions externally.
+func (o *BasicPodDefinitionManagerOptions) SetCache(pdc cocoa.ECSPodDefinitionCache) *BasicPodDefinitionManagerOptions {
+	o.Cache = pdc
+	return o
+}
+
+// SetCacheTag sets the tag used to track pod definitions in the cloud.
+func (o *BasicPodDefinitionManagerOptions) SetCacheTag(tag string) *BasicPodDefinitionManagerOptions {
+	o.CacheTag = &tag
+	return o
+}
+
+var (
+	defaultCacheTrackingTag = "cocoa-tracked"
+)
+
+// Validate checks that the required parameters to initialize a pod definition
+// manager are given.
+func (o *BasicPodDefinitionManagerOptions) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(o.Client == nil, "must specify a client")
+	catcher.NewWhen(o.CacheTag != nil && o.Cache == nil, "cannot specify a cache tracking tag when there is no cache")
+	if catcher.HasErrors() {
+		return catcher.Resolve()
+	}
+
+	if o.CacheTag == nil {
+		o.CacheTag = &defaultCacheTrackingTag
+	}
+
+	return nil
+}
+
+// NewBasicPodDefinitionManager creates a new pod definition manager optionally
+// backed by a cache.
+func NewBasicPodDefinitionManager(opts BasicPodDefinitionManagerOptions) (*BasicPodDefinitionManager, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid options")
+	}
+	return &BasicPodDefinitionManager{
+		client:   opts.Client,
+		vault:    opts.Vault,
+		cache:    opts.Cache,
+		cacheTag: utility.FromStringPtr(opts.CacheTag),
+	}, nil
+}
+
+// CreatePodDefinition creates a pod definition and caches it if it is using a
+// cache.
+func (m *BasicPodDefinitionManager) CreatePodDefinition(ctx context.Context, opts ...cocoa.ECSPodDefinitionOptions) (*cocoa.ECSPodDefinitionItem, error) {
+	mergedOpts := cocoa.MergeECSPodDefinitionOptions(opts...)
+	if err := mergedOpts.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid pod definition options")
+	}
+	if m.shouldCache() {
+		// If the definition needs to be cached, we could successfully create a
+		// cloud pod definition but fail to cache it. Adding a tag makes it
+		// possible to track whether the pod definition has been created but
+		// has not been successfully cached. In that case, the application can
+		// query ECS for pod definitions that are tagged as untracked to clean
```

Comment:
There's no fix for this issue, the thing I wrote in this ticket is the fix itself. It's on Evergreen the application to properly handle this tagging scheme.

Reply:
Yes I was referring to a ticket to handle this on the Evergreen side. If it doesn't exist in the current epic it may be worthwhile to make a ticket / TODO

Reply:
Ah, then the work to clean up the pod definitions using tags is tracked in [EVG-16900](https://jira.mongodb.org/browse/EVG-16900) (i.e. clean up those that we correctly cache after a TTL) and [EVG-16901](https://jira.mongodb.org/browse/EVG-16901) (i.e. clean up those that we fail to cache).